lib/reservation.rb

```diff
@@ -0,0 +1,90 @@
+require "date"
+require "csv"
+
+module Hotel
+  class Reservation
+    attr_reader :id, :room_ids, :start_date, :end_date, :block, :cost, :room_rate
+    attr_accessor :rooms
+    
+    CSV_RESERVATION = 'reservation.csv'
+    
+    def initialize(id:, rooms: nil, room_ids: nil, block: nil, start_date:, end_date:, room_rate: nil, cost: nil)
+      
+      validate_num(id)
+      @id = id
+      
+      if rooms
+        @rooms = rooms
+        @room_ids = @rooms.map {|room| room.id}
+        room_num = @rooms.length
+      elsif room_ids
+        @room_ids = room_ids
+        room_num = @room_ids.length
+      elsif block
+        @block = block
+        @room_ids = @block.map {|room| room.id}
+        room_num = @block.length
+      else
+        raise ArgumentError.new('Room or block info is required!')
+      end
+      
+      range = range(start_date, end_date)
+      @start_date = start_date
+      @end_date = end_date
+      
+      if cost
+        validate_num(cost)
+        @cost = cost
+      elsif room_rate
+        validate_num(room_rate)
+        @room_rate = room_rate
+        @cost = @room_rate * range * room_num
+      else
+        cost_per_night = 0
+        @room_ids.each do |room_id|
+          room = find_room(room_id)
+          cost_per_night += room.rate 
+        end
+        @cost = cost_per_night * range
+      end
+      
+      CSV.open(CSV_RESERVATION, 'a') do |csv|
+        csv_array = CSV.read(CSV_RESERVATION)
+        if block.nil?
+          block_reservation = 'No'
+        else
+          block_reservation = 'Yes'
+        end
+
+        new_csv = ["#{@id}", "#{@room_ids}", "#{block_reservation}", "#{@start_date}", "#{@end_date}", "#{room_rate}", "#{@cost}"]
+        
+        csv << new_csv if !csv_array.include?(new_csv)
+      end
+      
+    end
+    
+    def validate_num(num)
```

Comment:
There aren't any unit tests on this! There may be opportunity for this method to move to a different class, or to become a private method (or both)